monty-31/src/dft/backward.rs

```diff
@@ -0,0 +1,172 @@
+//! Discrete Fourier Transform, in-place, decimation-in-time
+//!
+//! Straightforward recusive algorithm, "unrolled" up to size 256.
+//!
+//! Inspired by Bernstein's djbfft: https://cr.yp.to/djbfft
+
+extern crate alloc;
+use alloc::vec::Vec;
+
+use super::split_at_mut_unchecked;
+use crate::{
+    monty_reduce, partial_monty_reduce, reduce_2p, reduce_4p, MontyField31, MontyParameters,
+    TwoAdicData,
+};
+
+impl<MP: MontyParameters + TwoAdicData> MontyField31<MP> {
+    #[inline(always)]
+    fn backward_butterfly(x: Self, y: Self, w: Self) -> (Self, Self) {
+        // TODO: See if doing a partial_monty_reduce followed by reduce_3p's is faster
+        let t = monty_reduce::<MP>(y.value as u64 * w.value as u64);
+        (
+            Self::new_monty(reduce_2p::<MP>(x.value + t)),
+            Self::new_monty(reduce_2p::<MP>(MP::PRIME + x.value - t)),
+        )
+    }
+
+    #[inline]
+    fn backward_pass(a: &mut [Self], roots: &[Self]) {
+        let half_n = a.len() / 2;
+        assert_eq!(roots.len(), half_n - 1);
+
+        let (top, tail) = unsafe { split_at_mut_unchecked(a, half_n) };
+
+        let x = top[0];
+        let y = tail[0];
+
+        let s = reduce_2p::<MP>(x.value + y.value);
+        let t = reduce_2p::<MP>(MP::PRIME + x.value - y.value);
+        top[0].value = s;
+        tail[0].value = t;
+
+        for i in 1..half_n {
+            (top[i], tail[i]) = Self::backward_butterfly(top[i], tail[i], roots[i - 1]);
+        }
+    }
+
+    #[inline(always)]
+    fn backward_2(a: &mut [Self]) {
+        assert_eq!(a.len(), 2);
+
+        let s = reduce_2p::<MP>(a[0].value + a[1].value);
+        let t = reduce_2p::<MP>(MP::PRIME + a[0].value - a[1].value);
+        a[0].value = s;
+        a[1].value = t;
+    }
+
+    #[inline(always)]
+    fn backward_4(a: &mut [Self]) {
+        assert_eq!(a.len(), 4);
+
+        // Read in bit-reversed order
+        let a0 = a[0].value;
+        let a2 = a[1].value;
+        let a1 = a[2].value;
+        let a3 = a[3].value;
+
+        let t1 = (MP::PRIME + a1 - a3) as u64;
+        let t5 = (a1 + a3) as u64;
+        let t3 = partial_monty_reduce::<MP>(t1 * MP::INV_ROOTS_8.as_ref()[1].value as u64) as u64;
+        let t4 = (a0 + a2) as u64;
+        let t2 = (MP::PRIME + a0 - a2) as u64;
+
+        a[0].value = reduce_4p::<MP>(t4 + t5);
+        a[1].value = reduce_4p::<MP>(t2 + t3);
+        a[2].value = reduce_4p::<MP>((2 * MP::PRIME as u64) + t4 - t5);
+        a[3].value = reduce_4p::<MP>((2 * MP::PRIME as u64) + t2 - t3);
+    }
+
+    #[inline(always)]
+    fn backward_8(a: &mut [Self]) {
+        assert_eq!(a.len(), 8);
+
+        let (a0, a1) = unsafe { split_at_mut_unchecked(a, a.len() / 2) };
```

Comment:
It feels slightly strange having both the assert_eq! and the unsafe. E.g. after the assertion the compiler knows that `split_at_mut(a, a.len() / 2)` will be valid.

(Maybe this is fixing a compiler error but if so might be worth adding a comment).

I'm just commenting this once but it applies to backward_16,32,... too

Reply:
Yeah, unfortunately the language doesn't let you omit the `unsafe`, even though the compiler knows it's safe after the `assert_eq()` (indeed, the assert is there principally so that the compiler will omit bounds checks). I'll add some comments to make this clear (in my mind the `assert_eq` was playing the role of such a comment, but you're right that it could be clearer still).